### ISSUE TITLE
infinite loop if cycle within cycle

### ISSUE THREAD

**Issue by yaoxinliu** [hanickadot/compile-time-regular-expressions#72]
The following expression will enter in an infinite loop! 
[Compiler: clang 8.0 with -std=c++2a]

```
R"(^(([^\\{}]*(\\.)?[^\\{}]*)*\{\}([^\\{}]*(\\.)?[^\\{}]*)*){2}$)"_ctre
        .match(R"(sds{}dsds{}ds)"sv);
```

error message:
```
./utl/ctre.hpp:3671:63: note: constexpr evaluation hit maximum step limit; possible infinite loop?
        constexpr CTRE_FORCE_INLINE void matched() noexcept { _matched = true; }
                                                              ^
./utl/ctre.hpp:3943:40: note: in call to '&captures._captures.head->matched()'
        _captures.template select<0>().matched();
                                       ^
./utl/ctre.hpp:4947:43: note: in call to '&captures->matched()'
    return captures.set_end_mark(current).matched();
```

**Comment by hanickadot** [comment]
it's not an infinite loop but heavy backtracking, the used algorithm will always ends, I would recommend to redesign the pattern

**Comment by yaoxinliu** [comment]
https://regex101.com works well on the same example, which takes less than 1ms.

**Comment by hanickadot** [comment]
Maybe other engines are doing some optimization for special cases, the regex you provide is the worst-case scenario for a backtracking engine...

`^(([^\\{}]*(\\.)?[^\\{}]*)*\{\}([^\\{}]*(\\.)?[^\\{}]*)*){2}$`

let's look at just the first part:

`([^\\{}]*(\\.)?[^\\{}]*)*`

which is basically:
- anything than backslash, open curly, close curly, zero or multiple time
- maybe a dot
- anything than backslash, open curly, close curly, zero or multiple time

and this whole is zero or multiple time


which means for the input `aaaaa` it can try to match:

nothing
only `a` (within first part)
only `a` (within second part)
only `aa` (within only first part)
only `aa` (within first and second part)
only `aa` (within only second part)

and we can add the dot somewhere:
`a.a` can be match by first part, dot part, and second part, but it can also be matched by only first part, as the dot can be matched with `[^\\{}]`

basically it tries all the combination, which explodes with the length of the input, hence the hitting constexpr limit of clang

I recommend add the dot to the `[^\\{}]` classes, so greedy => possessive optimization can help you. Or you can use possessive modifiers explicitly.

**Comment by hanickadot** [comment]
The thing is that CTRE doesn't contain any pattern based optimization with exception of greddy=>possessive transformation. It's pretty straightforward and it counts on compiler's optimizations. I will look into it deeply, but I'm not sure if I can help you significantly now.

**Comment by yaoxinliu** [comment]
Thank you very much for your detailed explanations. This is a great and excellent work!

**Comment by hanickadot** [comment]
Ok, there is a bug when I do the greedy cycle within greedy cycle, I minimized it into pattern:
`(a*)*`

I need to add forward guarantee in greedy cycle. Will need to think about it a little bit. Thank you for bringing this error.

**Comment by hanickadot** [comment]
It's going to be a bigger change than expected.

**Comment by hanickadot** [comment]
Fixed in branch recent release.
